CRM/Core/BAO/UFField.php

```diff
@@ -202,27 +257,39 @@ public static function checkMultiRecordFieldExists($gId) {
   public static function add(&$params) {
     // set values for uf field properties and save
     $ufField = new CRM_Core_DAO_UFField();
-    $ufField->copyValues($params);
-    $ufField->field_type = $params['field_name'][0];
-    $ufField->field_name = $params['field_name'][1];
 
-    //should not set location type id for Primary
-    $locationTypeId = NULL;
-    if ($params['field_name'][1] == 'url') {
-      $ufField->website_type_id = CRM_Utils_Array::value(2, $params['field_name']);
-    }
-    else {
-      $locationTypeId = CRM_Utils_Array::value(2, $params['field_name']);
-      $ufField->website_type_id = NULL;
-    }
-    if ($locationTypeId) {
-      $ufField->location_type_id = $locationTypeId;
-    }
-    else {
-      $ufField->location_type_id = 'null';
-    }
+    if (!is_string($params['field_name'])) {
+      $ufField->copyValues($params);
 
-    $ufField->phone_type_id = CRM_Utils_Array::value(3, $params['field_name'], 'NULL');
+      $field_type       = CRM_Utils_Array::value('field_type', $params);
+      $field_name       = CRM_Utils_Array::value('field_name', $params);
+      $location_type_id = CRM_Utils_Array::value('location_type_id', $params, CRM_Utils_Array::value('website_type_id', $params));
+      $phone_type       = CRM_Utils_Array::value('phone_type_id', $params, CRM_Utils_Array::value('phone_type', $params));
+
+      $params['field_name'] = array($field_type, $field_name, $location_type_id, $phone_type);
```

Comment:
Declaring 'field_name' here as an array feels kind of hack-ish and we can probably do without it. However since the parameter passed to this function was not a reference to the $params array, this should not cause any problems.